devmapper: Add a new option dm.min_free_space

rhvgoyal opened the pull request:
Fixes #20707 

Once thin pool gets full, bad things can happen. Especially in case of xfs
it is possible that xfs keeps on retrying IO infinitely (for certain kind
of IO) and container hangs. 

One way to mitigate the problem is that once thin pool is about to get full,
start failing some of the docker operations like pulling new images or
creation of new containers. That way user will get warning ahead of time
and can try to rectify it by creating more free space in thin pool. This
can be done either by deleting existing images/containers or by adding more
free space to thin pool.

This patch adds a new option dm.min_free_space_percent to devicemapper graph
driver. Say one specifies dm.min_free_space_percent=20. This means atleast
20% of data and metadata blocks should be free in pool before new device
creation is allowed, otherwise operation will fail.

By default min_free_space_percent is 20. User can change it by specifying
dm.min_free_space_percent=X on command line. A value of 0 will disable the
check.

Signed-off-by: Vivek Goyal vgoyal@redhat.com

rhvgoyal commented:
@samuelkarp  Give this patch a try and see if it helps.

thaJeztah commented:
I think this also addresses https://github.com/docker/docker/issues/20272

/cc @bfirsh

ChrisRut commented on docs/reference/commandline/daemon.md:
```diff
@@ -438,6 +438,31 @@ options for `zfs` start with `zfs`.
 
+*  `dm.min_free_space_percent`
+
+    Species the min free space percent in thin pool require for new device
```
s/Species/Specifies/g

ChrisRut commented on man/docker-daemon.8.md:
```diff
@@ -475,6 +475,29 @@ By default docker will pick up the zfs filesystem where docker graph
 
+#### dm.min_free_space_percent
+
+Species the min free space percent in thin pool require for new device
```
s/Species/Specifies/g

rhvgoyal replied:
Thanks. Will fix it.

rhatdan commented:
Should this be a percentage or a minimum free space?  Would saying I have to have 1G free in order to create another container make more sense, since 10% might have huge variant depending on the size of the device.

rhvgoyal commented:
I kind of like percentage more. One reason being that docker can have this as a in-built default value. Absolute values have to come from user.

Secondly we have precedence from lvm. We have automatic pool growing logic based on what percentage of space is free (and not absolute value). And also lvm has built in logic where new thin volume creation is denied if pool free space is below certain percentage threshold.

Agreed that percentage can create huge variation. 20% of 10G will be 2G while 20% of 100G will be 20G. But at the same time while 2G makes sense for 10G pool, I think 2G does not make much sense
for 100G pool. Very fact we have been able to fill up 98G so far, possibly our workload is more powerful and we are filling pool at faster rate and that means we need to warn little sooner then 2G free space.

So while percentage is not perfect, given it has precedence in lvm and given it is easy to use a percentage value as default, I prefer it.

rhvgoyal commented:
I think while percentage is not perfect, it works for pools of all size. And if this becomes a problem we can always think of implementing another parameter which can override this and enforce absolute values. Say dm.min_free_space=XG

thaJeztah commented:
Or have `dm.min_free_space` accept both `20%` and `20G`? Having both `min_free_space_percent` and `min_free_space` might bring more confusion (i.e. what would happen if I specify both)

(thinking out loud again :smile:)

rhvgoyal commented:
If one specifies both, then either we can error out or define that absolute value trumps percentage value or define one defined in the end of command line takes affect. 

I would prefer to keep them as two separate parameters. But if there is precedence somewhere else in the code, it makes your case little stronger.

rhvgoyal commented:
IOW, I have not seen parameters taking both percentage and absolute values as input and that's the reason I am hesitant to create one like that.

rhatdan commented:
Well I don't have great arguments against percentage, just wanted to bring it up for discussion.

redbaron commented:
> IOW, I have not seen parameters taking both percentage and absolute values as input and that's the reason I am hesitant to create one like that

lvchange/lvcreate is a perfect example of that,where '-l' accepts both percentage (%FREE, %VG, %PVS) and absolute number of extents which is _very_ convenient and works wonders.

rhvgoyal commented:
@redbaron, that's good point. I guess I will look into supporting both % and absolute value options using same parameter. dm.min_free_space.

rhvgoyal commented:
Thinking more about it, this percentage limit applies both to data as well metadata. So if we have 20% free data space but not 20% free metadata space, then also error will occur. 

If we start specifying absolute values, then we will have to have separate knobs for data and metadata as one value for both does not make sense (as typically metadata is very small).

Say dm.min_free_data=2GB and dm.min_free_metadata=2MB

Notion of percentage works well both for data and metadata and no separate knobs are required. 

So I am not writing another patch to support both percentage and absolute value using same option. Leaving patch as it is.

rhvgoyal commented:
Pushed patch again. Fixed the spelling typos in documentation. Also tweaked error message a bit if pool does not have sufficient space.

thaJeztah commented:
```
16:19:42 /go/src/github.com/docker/docker/hack/make/test-integration-cli: line 18: 11665 Killed                  ( bundle .integration-daemon-start; bundle .integration-daemon-setup; bundle_test_integration_cli; bundle .integration-daemon-stop ) 2>&1
16:19:42      11666                       | tee -a "$DEST/test.log"
16:19:46 Build step 'Execute shell' marked build as failure
```
https://jenkins.dockerproject.org/job/Docker-PRs/24433/console

rhvgoyal commented:
@thaJeztah Not sure what the problem is. Almost same patch passed last time. I don't think this failure is due to my patch.

thaJeztah commented:
@rhvgoyal oh, sorry, no it's not related; were still investigating this one, as we have no clue what kills the tests

Fredjet commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  fredjet15@gmail.com
    Domain ventureacceleration.com has exceeded the max emails per hour (58/50 (115%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext3.iad.github.net ([192.30.252.194]:53639 helo=github-smtp2b-ext-cp1-prd.iad.github.net)
	by eaglepeak.ovasp.com with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.86)
	(envelope-from <noreply@github.com>)
	id 1aarmM-0008BP-Cb
	for fred@ventureacceleration.com; Tue, 01 Mar 2016 16:24:27 -0500
Date: Tue, 01 Mar 2016 10:23:31 -0800
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1456856611;
	bh=Ho3896NPIYod/kWhQWLROKCFsWvuXU6JGsK67+i9GXA=;
	h=From:Reply-To:To:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=tzKtc1pfsN79lbG2fTnfA16ZfeS/pBaJ2IrJx+R4kk5jTSo+LgBuXJ8Mp7WcWS2WF
	 Qkze3EiROwIB4azalANgCDgKT5JfQBA1CMTVyTOC7BjlFOj2CeiBmYxC4ryd/XkAh+
	 QEamJsMZITuMoPx2FikaJeSeJhbR/RzG7wDymQLk=
From: Vivek Goyal <notifications@github.com>
Reply-To: docker/docker <reply@reply.github.com>
To: docker/docker <docker@noreply.github.com>
Message-ID: <docker/docker/pull/20786/c190841358@github.com>
In-Reply-To: <docker/docker/pull/20786@github.com>
References: <docker/docker/pull/20786@github.com>
Subject: Re: [docker] devmapper: Add a new option dm.min_free_space_percent
 (#20786)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_56d5de23be4e9_96f3f966b81b2bc759b7";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: rhvgoyal
X-GitHub-Recipient: Fredjet
List-ID: docker/docker <docker.docker.github.com>
List-Archive: https://github.com/docker/docker
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+0007d4122b9070f32b70ac3d7110f96ee2e72b8c03a4714892cf0000000112eda02392a169ce082ecb90@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AAfUEhoS5Cm_j0UnpPm00efLHb1zM2Ywks5ppIQjgaJpZM4HlgwK>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: fred@ventureacceleration.com


----==_mimepart_56d5de23be4e9_96f3f966b81b2bc759b7
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

@thaJeztah Not sure what the problem is. Almost same patch passed last time. I don't think this failure is due to my patch.

---
Reply to this email directly or view it on GitHub:
https://github.com/docker/docker/pull/20786#issuecomment-190841358
----==_mimepart_56d5de23be4e9_96f3f966b81b2bc759b7
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<p><a href="https://github.com/thaJeztah" class="user-mention">@thaJeztah</a> Not sure what the problem is. Almost same patch passed last time. I don't think this failure is due to my patch.</p>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br>Reply to this email directly or <a href="https://github.com/docker/docker/pull/20786#issuecomment-190841358">view it on GitHub</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/AAfUEuGfFKfWgeIAfC04fuLOfPOoysK-ks5ppIQjgaJpZM4HlgwK.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
<div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
  <link itemprop="url" href="https://github.com/docker/docker/pull/20786#issuecomment-190841358"></link>
  <meta itemprop="name" content="View Pull Request"></meta>
</div>
<meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

----==_mimepart_56d5de23be4e9_96f3f966b81b2bc759b7--

rhatdan commented:
I think the default for this should probably be 90% which is the default for a lot of file systems, where I believe you need to have sys_resource capability (root) to add content to the file system.

rhvgoyal commented:
Ok, I had just picked a random number for default. I like 90% better (instead of 80%). Will change it.

thaJeztah commented:
Maintainers meeting review on design:

We're ok with just doing percentages for now, and only implement absolute values (GB) if there's a real need for that. *However* we would like it to be a single option, not separate options for "percent" and "absolute value", so:

- Strip the `_percent` from the option-name
- Values *must* be specified with `%` (`--storage-opt dm.min_free_space=10%`)

We really like this option, so thanks for implementing!

rhvgoyal commented:
@thaJeztah ok, will implement the two suggestions you made and generate PR again.

rhvgoyal commented:
@thaJeztah Pushed a new patch. Took care of your comments and also changed default value of min_free_space to 10%.

thaJeztah commented on docs/reference/commandline/daemon.md:
```diff
@@ -438,6 +438,32 @@ options for `zfs` start with `zfs`.
+
+    Example use:
+
+     $ docker daemon --storage-opt dm.min_free_space_percent=10%
```
oh, my eye fell on this (not in docs review yet); you need to increase the indentation to 8 spaces total here (because it's in a list), otherwise it isn't rendered as a code-block

rhvgoyal replied:
Got it. Will fix it.

thaJeztah commented:
Thanks!

ping @vbatts @calavera could you have a look at the code changes?

calavera commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -2512,6 +2551,22 @@ func NewDeviceSet(root string, doInit bool, options []string, uidMaps, gidMaps [
+			}
+
+			if minFreeSpacePercent >= 100 {
+				return nil, fmt.Errorf("devmapper: Invalid value %v for optiona dm.min_free_space", val)
```
`s/optiona/option/`

rhvgoyal replied:
thanks. will fix.

rhvgoyal committed:
devmapper: Add a new option dm.min_free_space

Once thin pool gets full, bad things can happen. Especially in case of xfs
it is possible that xfs keeps on retrying IO infinitely (for certain kind
of IO) and container hangs. 

One way to mitigate the problem is that once thin pool is about to get full,
start failing some of the docker operations like pulling new images or
creation of new containers. That way user will get warning ahead of time
and can try to rectify it by creating more free space in thin pool. This
can be done either by deleting existing images/containers or by adding more
free space to thin pool.

This patch adds a new option dm.min_free_space to devicemapper graph
driver. Say one specifies dm.min_free_space=10%. This means atleast
10% of data and metadata blocks should be free in pool before new device
creation is allowed, otherwise operation will fail.

By default min_free_space is 10%. User can change it by specifying
dm.min_free_space=X% on command line. A value of 0% will disable the
check.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
Took care of review comments and pushed new patch.

thaJeztah commented:
Userns failure is a configuration error on CI in the post build step https://jenkins.dockerproject.org/job/Docker-PRs-userns/7042/console;

```
21:06:53 tar: unrecognized option '--include'
21:06:53 Try 'tar --help' or 'tar --usage' for more information.
```

WindowsTP4 is a network issue. I think that node has been restarted since, so I'll re-run

``
21:45:20 fatal: unable to access 'https://github.com/docker/docker.git/': Couldn't resolve host 'github.com'
```

vbatts commented:
SGTM. Looking over the the design and code, look good. I have not tested though.

rhvgoyal commented:
@thaJeztah Ping. How can I make progress on this one.

thaJeztah commented:
Let me rattle some cages

ping @calavera @vbatts PTAL

calavera commented:
LGTM

vbatts commented:
LGTM
